User and Read-only API keys

tinahollygb opened the pull request:
## Features and Changes

This PR adds support for User and Read-only API keys.

It supersedes https://github.com/growthbook/growthbook/pull/1258

- Closes https://linear.app/growthbook/issue/GB-192
- Closes https://linear.app/growthbook/issue/GB-190
- Closes https://linear.app/growthbook/issue/GB-201
- Closes https://linear.app/growthbook/issue/GB-202
- Closes https://linear.app/growthbook/issue/GB-181
- Closes https://linear.app/growthbook/issue/GB-180

### Data modelling and auth middleware checks

Adds fields to the ApiKey model, and some helper methods for creating read-only and user API tokens.

Also adds the check in the auth middleware and some unit tests.


### API key management UI

Allows the user to create and delete API keys of various types:

- Secret keys: these are the existing ones with full access. They are evaluated to have the role `admin`.
- Personal Access Tokens (user API keys): these would check their user account to verify permissions and evaluate the role to be the current user's role provided they're still in the org
- Read-only keys: these would allow only read-only access and have an assigned role of `readonly` when inserted in the DB.


## Dependencies

n/a

### Testing

You can also run the app, navigate to http://localhost:3000/settings/keys and create read-only and secret keys as demonstrated in the attached video. Create and reveal them.

Then, navigate to the My Personal Access Tokens page from the user menu and create and reveal personal access tokens.

### Screenshots

<img width="347" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/1f0b8745-47a5-4aff-bf27-991b4cba0dfc">

Unified table for secret API keys with role column:

<img width="1436" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/1d7403fe-971d-45cf-850e-e40d9ef00763">


Personal access tokens for user:

<img width="1440" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/c9c05317-8695-4621-bfc5-e582cff3687c">





The UI in this video is outdated since the user API keys have been renamed to personal access tokens and have moved to their own page, but it's more or less the same.

https://user-images.githubusercontent.com/113377031/236937177-0437883f-39c2-467c-ae90-731cf620785b.mp4

bryce-fitzsimons commented on packages/front-end/components/Settings/SecretApiKeys.tsx:
```diff
@@ -14,103 +15,202 @@ const SecretApiKeys: FC<{ keys: ApiKeyInterface[]; mutate: () => void }> = ({
+  onDelete: (keyId: string | undefined) => () => Promise<void>;
+  keys: ApiKeyInterface[];
+  canManageKeys: boolean;
+  onReveal: (keyId: string | undefined) => () => Promise<string>;
```
consider a `maxHeight` and `overflow: auto` since we're stacking these vertically and each table could potentially get quite long

bryce-fitzsimons reviewed (commented):
I'm likely missing something obvious here, but when I click "Create New User Key", it goes through the `postApiKey` flow and never actually assigns a userId. My data model for the key does not have the userId either. Is this intentional, oversight, or did I do something wrong?

Follow up question: are User API Keys supposed to show any information about which user they belong to? Can an admin set the user? Is a non-admin user supposed to be able to create their own keys (as it doesn't seem they can even reach /settings/keys)? More broadly, I guess I don't fully understand the use case for user keys.

tinahollygb commented:
Thanks @bryce-fitzsimons for the thorough review. Looks like an oversight re: forgetting to assign the `userId` to the API key. This has been fixed here: 014f6a28d3828bd63873e575728dc971dc4e82e4

bryce-fitzsimons commented on packages/back-end/src/routers/organizations/organizations.controller.ts:
```diff
@@ -1183,9 +1183,13 @@ export async function putOrganization(
 export async function getApiKeys(req: AuthRequest, res: Response) {
   const { org } = getOrgFromReq(req);
   const keys = await getAllApiKeysByOrganization(org.id);
+  const filteredKeys = keys.filter(
```
Slightly confused on the UX here.
- Shouldn't an admin user be able to manage all keys?
- If so, should we also show which user owns each key?
- Is there any way to generate a key for a non-admin (as the UI doesn't allow non-admins to visit this page)?

tinahollygb replied:
Regarding your first 2 questions, I think they're answered here: https://github.com/growthbook/growthbook/pull/1265#issuecomment-1548414344

Regarding the 3rd, not at this time, but they can generate a read-only token and share it with unprivileged users.

tinahollygb commented:
I've added a change here 92f6b929edd51528d5a9fc0d3f11c97143042227 to hide user API keys unless they're your own. So if I create a user API key from my 2nd account, my 1st account shouldn't see it.

@bryce-fitzsimons re: this question:

> Follow up question: are User API Keys supposed to show any information about which user they belong to? 

Currently only the user ID is attached to the key and it is not visible in the UI. On the API Keys page, only your own user API keys are returned.


> Can an admin set the user? 

No. The API keys are created using the authenticated user ID and are not passed in via the UI. This would likely be a security concern, e.g. an admin could create keys on behalf of another user and use that key, which is essentially impersonating them with any actions performed. I don't think we should support that.

> Is a non-admin user supposed to be able to create their own keys (as it doesn't seem they can even reach /settings/keys)? 

Only users with the permission `manageApiKeys` should be able to access it.

This actually made me think of another security issue, which I've fixed here: 20fea25f8f81befba945f2b1c0a675442d19b838

> More broadly, I guess I don't fully understand the use case for user keys.

Currently the public API has no permission checks. All API keys grant full access to all resources. This is unsuitable for most use cases. Many organizations want more finer-grained controls over permissions. The intention of user API keys is to allow all permissions a user would have in the app.

bryce-fitzsimons reviewed (approved):
Looks good after feedback / questions addressed.

I do think it's useful (almost expected?) for an admin user to be able to see all user keys (not just the ones they have created). But I don't want this to be a blocker.

jdorn commented on packages/back-end/src/api/openapi/openapi.yaml:
```diff
@@ -13,7 +13,11 @@ info:
+    
+    - **Secret keys**: These are sensitive and provide full access to the organization
+    - **User API keys**: These are sensitive and provide the same level of access as the user has to an organization
+    - **Read-only keys**: These are sensitive and provide read access to resources under an organization
```
I think we should only have 2 types of API keys

1. **Personal Access Tokens** which have a `userId` and lookup the role in the organization.  Any user manage their own Personal Access Tokens.
2. **Secret Keys** which have an explicit `role` specified, but no `userId`.  Only Admins with the `manageApiKeys` permission can manage Secret Keys.

For existing secret API keys, they would fall under 2 above and would default to `role = "admin"` to match the existing behavior.  Admins can create new keys with `role = "readonly"` if they want (or any other role).

This approach will be less confusing to users, more flexible, and will have less hard-coded permission logic (e.g. if type = "read-only", throw an error).  With this change, every incoming request will have a role associated with it, so you can always just check if the role contains the permission.

tinahollygb replied:
I've made changes around supporting a role property for ApiTokens. Please re-review.

jdorn commented on packages/front-end/components/Settings/SecretApiKeys.tsx:
```diff
@@ -14,103 +15,202 @@ const SecretApiKeys: FC<{ keys: ApiKeyInterface[]; mutate: () => void }> = ({
+
+      {/* region User API keys */}
+      <div className="mb-5">
+        <h1>User API Keys</h1>
```
The UI doesn't quite feel right here.  I'd expect every user to be able to manage their own Personal Access Tokens, even if they are not an admin and missing the `manageApiKeys` permission.  As such, I'd expect it to be accessible from the user dropdown in the top nav, not under Settings.

The **Settings -> API Keys** page will be just for admins and contain only the secret keys.  I think this will be much less confusing for users.

tinahollygb replied:
As suggested, I've added the personal access tokens to a new page (under settings but not in the settings menu) under the user menu.

tinahollygb commented on packages/back-end/src/services/vercel.ts:
```diff
@@ -65,15 +65,15 @@ export async function createOrgGbKeys(
-      environment: envMapEntry.gb,
-      project: "",
-      encryptSDK: false,
+    const createdKeyVal = await createLegacySdkKey({
+      organizationId: orgId,
       description:
         "This key is used by Vercel that allows you to connect your GrowthBook sdk to the GrowthBook API.",
+      project: "",
+      encryptSDK: false,
+      environment: envMapEntry.gb,
```
Let me know if this Vercel integration key is ok @jdorn

jdorn commented on packages/back-end/src/models/ApiKeyModel.ts:
```diff
@@ -69,26 +85,108 @@ export function generateSigningKey(prefix: string = "", bytes = 32): string {
+  organizationId: string;
+  description: string;
+  environment: string;
+  project: string;
```
We don't need `environment` and `project` for these.  That's only for the legacy SDK keys

jdorn commented on packages/back-end/src/models/ApiKeyModel.ts:
```diff
@@ -69,26 +85,108 @@ export function generateSigningKey(prefix: string = "", bytes = 32): string {
-  const prefix = secret ? "secret_" : `${getShortEnvName(environment)}_`;
+  const prefix = secret
+    ? "secret_" + (role ? `${role}_` : "")
+    : `${getShortEnvName(environment)}_`;
```
This will make user access tokens look like `secret_...`, which is identical to the existing admin secret tokens in Mongo.  Should we try to differentiate these so it's easy to tell if you're using a personal key vs an organization one?

tinahollygb replied:
I pushed up a change to support this. c45c6ba26ff5cafb4e0fe824301dac8110fcb0f9

<img width="426" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/d781125a-4258-455e-b6c2-8cd58b9e4f9c">

<img width="375" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/5d335ad0-7273-4c90-a8dd-0738e4c7f2a4">

jdorn commented on packages/front-end/components/Layout/TopNav.tsx:
```diff
@@ -227,6 +227,17 @@ const TopNav: FC<{
                 </>
               )}
               <div className="dropdown-divider"></div>
+              <Link href={"/settings/personal-access-tokens"}>
```
The `/settings/` path today is specifically for admin-only settings.  Maybe we should use a new top-level path like `/account/`?

tinahollygb replied:
moved to `/account`

jdorn commented on packages/front-end/components/PersonalAccessTokens/PersonalAccessTokens.tsx:
```diff
@@ -0,0 +1,123 @@
+        <p>
+          User keys have full read and write access to your user account.
+          Because of this, they must be kept secure and{" "}
+          <strong>must not be exposed to users</strong>.
```
Is the use of `users` here confusing?  "user tokens must not be exposed to users"

tinahollygb replied:
updated the copy to make it less redundant

jdorn commented on packages/front-end/services/secret-api-keys.utils.ts:
```diff
@@ -0,0 +1,38 @@
+ *  - Read-only keys have type 'read-only'
+ * @param apiKeys List of {@link ApiKeyInterface}
+ */
+export function groupApiKeysByType(apiKeys: ApiKeyInterface[]): GroupedApiKeys {
```
Is this used any more?

tinahollygb replied:
It is. If you find on the PR/files page you should see it [here](https://github.com/growthbook/growthbook/pull/1265/files#diff-e53ac8e3bfd54803b3be3e2acd31a9a0e44f45333f04620cc1c2183085816cceR79).

tinahollygb commented on packages/back-end/src/middleware/authenticateApiRequestMiddleware.ts:
```diff
@@ -86,3 +126,87 @@ export default function authencateApiRequestMiddleware(
+  organization,
+  environments,
+  project,
+}: VerifyApiKeyPermissionOptions) {
```
I made the selected change, see below.

tinahollygb commented on packages/back-end/src/routers/organizations/organizations.controller.ts:
```diff
@@ -1218,17 +1230,60 @@ export async function postApiKey(
-    secret: !!secret,
+  // Handle user personal access tokens
+  if (type === "user") {
+    if (!req.userId) {
```
8ad7e5677e35551725491ccbc6f2eb0bddfc193d

tinahollygb replied:
It looks like that's just an abstraction that uses `req.userId` but i made the change since it seems that's the preferred way to do it. Based on the name of the method, it didn't look like it would return the userId and only the organization.

tinahollygb commented on packages/front-end/components/Settings/ApiKeysModal.tsx:
```diff
@@ -100,6 +112,23 @@ const ApiKeysModal: FC<{
           setValue={(value) => form.setValue("encryptSDK", value)}
         />
       )}
+      {type !== "user" && (
```
added
